Add dynamic selection between `memset` and `parallel_for` for `View` initialization in the OpenMP backend

Adrien-Tab opened the pull request:
This PR introduces dynamic selection between `memset` and `parallel_for` for memory initialization for OpenMP and Threads backend.

### Changes
- Three new performance tests for rank-1 views.
- Each host backend has its own `ZeroMemset` function
- Removed `core/src/impl/Kokkos_HostSpace_ZeroMemset.hpp`

The method switches at runtime the way to initialize the `View` with a threshold set at **2^17** bytes.

**Performance results for AMD Genoa EPYC 9654 :**
![adastra_switch_17_initialize](https://github.com/user-attachments/assets/c1f9f8dd-2e5d-485c-bcf6-f312ec6ad721)
![adastra_switch_17_parallel_for](https://github.com/user-attachments/assets/f49288f4-08ce-410a-b7ad-6703667936a9)

**Performance results for Intel Xeon Gold 6230 :**
![ruche_switch_17_initialize](https://github.com/user-attachments/assets/6a6cfc23-7272-4238-ba0f-856f92a84fc4)
![ruche_switch_17_parallel_for](https://github.com/user-attachments/assets/e4595eb1-fc8f-4184-95f6-9e21267503ea)

Adrien-Tab committed:
No memset for OpenMP

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab committed:
Add OpenMP ZeroMemset

Switch OpenMP

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab committed:
Adjusts minimum for parallel_for to 2^17

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab committed:
Reduce Test size

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

JBludau commented on core/src/impl/Kokkos_HostSpace_ZeroMemset.hpp:
```diff
@@ -24,6 +24,7 @@
 namespace Kokkos {
 namespace Impl {
 
+#ifndef KOKKOS_ENABLE_OPENMP
```
Hmm ... not liking he ifdef here ... but I don't see a good way right now. Could you at least add a comment saying that OpenMP has its own?
Or think about how to push this into the host backend impls since HPX will need the fence, OpenMP (and probably Threads) would get the `memset` vs `parallel_for`, and Serial is fine with just memset ...

Adrien-Tab replied:
I have to use an `#ifdef` here to avoid instantiating the template. I didn’t find a better solution, so I’ll leave it like this with a comment.

JBludau replied:
yeah, I know. The idea was to ditch this header and only have the versions in the host-backend-specific files

masterleinad replied:
That would require a full definition of the execution space which we might or might not have here.

JBludau replied:
@AdRi1t could you check?

Adrien-Tab replied:
yes

Adrien-Tab committed:
Add a comment

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

masterleinad commented on core/perf_test/PerfTest_ViewFirstTouch.hpp:
I'd rather split up this file and inlcude the respective chunks in the 3 `*.cpp` files.

Adrien-Tab replied:
I think these tests should be separated into another executable (they are currently in `PerformanceTest_Benchmark`) to avoid being timed out by Google Test.

masterleinad reviewed (commented):
Otherwise, this looks good to me.

Adrien-Tab committed:
ZeroMemset for each Host backends

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

JBludau commented on core/src/Threads/Kokkos_Threads_ZeroMemset.hpp:
```diff
@@ -0,0 +1,46 @@
+namespace Kokkos {
+namespace Impl {
+
+template <>
+struct ZeroMemset<Threads> {
+  ZeroMemset(const Threads& exec_space, void* dst, size_t cnt) {
+    if (cnt < 0x20000ul) {  // 2^17
+      std::memset(dst, 0, cnt);
+    } else {
+      Kokkos::parallel_for(
+          "Kokkos::ZeroMemset via parallel_for",
+          Kokkos::RangePolicy<Kokkos::Threads, Kokkos::IndexType<size_t>>(
+              exec_space, 0, cnt),
+          KOKKOS_LAMBDA(size_t i) { static_cast<char*>(dst)[i] = 0; });
+    }
+  }
+};
```
This is a repetition of the OpenMP implementation. Did you check that this has the same perf results?

Adrien-Tab replied:
> This is a repetition of the OpenMP implementation. Did you check that this has the same perf results?

Yes, I ran some tests, and it looks like the same issue as with OpenMP.

**Threads Backends + gcc 13.2**

**Performance results for AMD Genoa EPYC 9654 :**
![adastra_192_threads_initilaize](https://github.com/user-attachments/assets/96e0fe10-28fb-49b0-83ac-08d54af0dd43)
![adastra_192_threads_parallel_for](https://github.com/user-attachments/assets/9bcc6540-f12e-4b35-b692-4a28bea72c7a)

**Performance results for Intel Xeon Gold 6230 :**
![ruche_40_Initialize_threads](https://github.com/user-attachments/assets/01f942b9-4e0f-48b6-af59-225c69c0c6da)
![ruche_40_parallel_for_threads](https://github.com/user-attachments/assets/1017454c-550b-4bb9-85cc-aa40924edc89)

Adrien-Tab committed:
Correction for HPX and CI

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab committed:
Removing unnecessary template instantiations

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

masterleinad commented on core/src/Threads/Kokkos_Threads_ZeroMemset.hpp:
```diff
@@ -0,0 +1,43 @@
+template <>
+struct ZeroMemset<Threads> {
+  ZeroMemset(const Threads& exec_space, void* dst, size_t cnt) {
+    constexpr size_t host_memset_limit = 0x20000ul;  // 2^17
```
```suggestion
    constexpr size_t host_memset_limit = (1lu << 17);
```

masterleinad commented on core/src/OpenMP/Kokkos_OpenMP_ZeroMemset.hpp:
```diff
@@ -14,29 +14,30 @@
-    std::memset(dst, 0, cnt);
+struct ZeroMemset<OpenMP> {
+  ZeroMemset(const OpenMP& exec_space, void* dst, size_t cnt) {
+    constexpr size_t host_memset_limit = 0x20000ul;  // 2^17
```
```suggestion
    constexpr size_t host_memset_limit = 1lu<<17;
```

masterleinad commented on core/src/HPX/Kokkos_HPX.hpp:
```diff
@@ -528,6 +528,14 @@ struct MemorySpaceAccess<Kokkos::Experimental::HPX::memory_space,
+template <>
+struct ZeroMemset<Kokkos::Experimental::HPX> {
+  ZeroMemset(const Kokkos::Experimental::HPX &exec, void *dst, size_t cnt) {
+    exec.fence("ZeroMemset");
```
```suggestion
    exec.fence("Kokkos::Impl::ZeroMemset: fence before calling std::memset");
```

masterleinad reviewed (commented):
How long do the new benchmarks run for?

Adrien-Tab committed:
Change fence name HPX

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab committed:
Update core/src/Threads/Kokkos_Threads_ZeroMemset.hpp

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

Adrien-Tab committed:
Update core/src/OpenMP/Kokkos_OpenMP_ZeroMemset.hpp

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

Adrien-Tab committed:
Apply clang format

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

dalg24 commented on core/src/Threads/Kokkos_Threads_ZeroMemset.hpp:
```diff
@@ -0,0 +1,43 @@
+
+#include <Kokkos_Macros.hpp>
+#include <Threads/Kokkos_Threads.hpp>
+#include <Threads/Kokkos_Threads_Instance.hpp>
```
What do we need this header include for?

dalg24 replied:
```C++
#include <cstring>  // memset

dalg24 commented on core/src/OpenMP/Kokkos_OpenMP_ZeroMemset.hpp:
```diff
@@ -14,29 +14,30 @@
 #include <Kokkos_Macros.hpp>
-#include <Kokkos_HostSpace.hpp>
+#include <OpenMP/Kokkos_OpenMP.hpp>
+#include <OpenMP/Kokkos_OpenMP_Instance.hpp>
```
What do we need this header include for?

dalg24 commented on core/src/Kokkos_CopyViews.hpp:
```diff
@@ -26,7 +26,7 @@ static_assert(false,
 #include <KokkosExp_MDRangePolicy.hpp>
 #include <Kokkos_Layout.hpp>
-#include <impl/Kokkos_HostSpace_ZeroMemset.hpp>
+#include <impl/Kokkos_ZeroMemset_fwd.hpp>
```
I think that header is not sufficient but maybe we can defer cleaning that up to later.

masterleinad replied:
I'm not sure it's even worth it having the file for a single forward declaration.

dalg24 replied:
Not sure what you are saying

dalg24 commented on core/src/HPX/Kokkos_HPX.hpp:
```diff
@@ -528,6 +528,15 @@ struct MemorySpaceAccess<Kokkos::Experimental::HPX::memory_space,
   enum : bool { deepcopy = false };
 };
 
+template <>
+struct ZeroMemset<Kokkos::Experimental::HPX> {
+  ZeroMemset(const Kokkos::Experimental::HPX &exec, void *dst, size_t cnt) {
+    exec.fence(
+        "Kokkos::Impl::ZeroMemset: HostSpace fence before calling std::memset");
+    std::memset(dst, 0, cnt);
+  }
+};
```
Fine.
Just making sure @msimberg sees this.

dalg24 reviewed (commented):
The runtime decision probably also needs to factor in the number of threads

Adrien-Tab committed:
Fix includes

- No benchmark added to tests in debug mode
- Takes concurrency into account in algorithm selection

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

romintomasetti commented on core/perf_test/CMakeLists.txt:
```diff
@@ -70,8 +70,9 @@ function(KOKKOS_ADD_BENCHMARK NAME)
   set(BENCHMARK_ARGS --benchmark_counters_tabular=true --benchmark_out=${BENCHMARK_NAME}_${BENCHMARK_TIME}.json)
-
-  add_test(NAME ${BENCHMARK_NAME} COMMAND ${BENCHMARK_NAME} ${BENCHMARK_ARGS})
+  if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
+    add_test(NAME ${BENCHMARK_NAME} COMMAND ${BENCHMARK_NAME} ${BENCHMARK_ARGS})
+  endif()
```
Shouldn't this be in its own, identifiable PR ?

dalg24 replied:
Agreed.  Please revert this change, it does not belong to this PR.

Adrien-Tab replied:
done

dalg24 commented on core/perf_test/PerfTest_ViewFirstTouch_Initialize.cpp:
```diff
@@ -0,0 +1,54 @@
+  using ViewType = Kokkos::View<DataType*>;
+
+  for (auto _ : state) {
+    Kokkos::fence();
```
I don't think this fence is needed

dalg24 commented on core/perf_test/PerfTest_ViewFirstTouch_ParallelFor.cpp:
```diff
@@ -0,0 +1,59 @@
+  }
+}
+
+namespace Test {
```
Either drop the namespace or enclose the function template and rename Test -> Benchmark.

dalg24 commented on core/perf_test/PerfTest_ViewFirstTouch_DeepCopy.cpp:
```diff
@@ -0,0 +1,56 @@
+  for (auto _ : state) {
+    Kokkos::fence();
+
+    ViewType v_a("A", N);
```
I would hoist the allocation out of the loop.

Adrien-Tab committed:
Revert change in perf_test/CMakeLists.txt and update hostspace_parallel_zeromemset function

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab commented:
> How long do the new benchmarks run for?

New benchmarks take about 130 seconds to run.

dalg24 commented on core/src/impl/Kokkos_HostSpace_deepcopy.cpp:
```diff
@@ -137,10 +137,51 @@ void hostspace_parallel_deepcopy_async(const ExecutionSpace& exec, void* dst,
+    *dst_c = z_u8;
+    dst_c++;
+  }
+#if (defined(KOKKOS_ENABLE_HPX) && \
+     defined(KOKKOS_ENABLE_IMPL_HPX_ASYNC_DISPATCH))
+  exec.fence(
+      "Kokkos::Impl::hostspace_parallel_zeromemset: fence after zeromemset");
+#endif
```
Does it belong here

Adrien-Tab replied:
I removed the fence. I changed to call one kernel for the HPX backend, nothing is done synchronously.

Adrien-Tab committed:
Sequence the kernel correctly for the HPX backend

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab committed:
Reduces the number of benchmarks

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

Adrien-Tab committed:
fix unused variable

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

JBludau commented on core/perf_test/PerfTest_ViewFirstTouch_DeepCopy.cpp:
```diff
@@ -0,0 +1,63 @@
+    ->Ranges({{int64_t(1) << 6, int64_t(1) << 24}, {0, 1}})
+    ->UseManualTime();
+
+#if 0  // Larger benchmark set
```
Is that an oversight?

Adrien-Tab replied:
No, that's intentional. It's for those who want to run the full benchmark.

masterleinad replied:
How expensive is that in `Release` mode? Dead code doesn't really help. Either we test it or we remove it.

Adrien-Tab replied:
I’ve removed the dead code and kept the small benchmarks — they take around 30 seconds to run in Release mode.

JBludau commented on core/perf_test/PerfTest_ViewFirstTouch_Initialize.cpp:
```diff
@@ -0,0 +1,61 @@
+    ->Range(int64_t(1) << 6, int64_t(1) << 24)
+    ->UseManualTime();
+
+#if 0  // Larger benchmark set
```
same here

JBludau commented on core/perf_test/PerfTest_ViewFirstTouch_ParallelFor.cpp:
```diff
@@ -0,0 +1,68 @@
+namespace Benchmark {
+
+template <typename DataType>
+void ViewFirstTouch_ParallelFor(benchmark::State& state) {
```
Could you explain how you got to the naming? I have trouble seeing the connection between FirstTouch and what you are doing here

Adrien-Tab replied:
This test is meant to measure the potential impact of first-touch memory placement on a `parallel_for` loop.

dalg24 replied:
Jakob are you ok with the name or do you want to suggest another better name?

JBludau replied:
I am fine with keeping it

JBludau commented on core/src/OpenMP/Kokkos_OpenMP_ZeroMemset.hpp:
```diff
@@ -14,29 +14,30 @@
+#include <OpenMP/Kokkos_OpenMP.hpp>
 #include <impl/Kokkos_ZeroMemset_fwd.hpp>
 
+#include <cstring>
```
why do we need it here?

Adrien-Tab replied:
We need it here because `memset` is defined in `<cstring>`.

JBludau commented on core/src/Threads/Kokkos_Threads_ZeroMemset.hpp:
```diff
@@ -0,0 +1,43 @@
+struct ZeroMemset<Threads> {
+  ZeroMemset(const Threads& exec_space, void* dst, size_t cnt) {
+    constexpr size_t host_memset_limit = 1lu << 17;
+    if (cnt < host_memset_limit || exec_space.concurrency() < 4) {
```
Can you add a comment which benchmarks were run to determine these limits on which hardware, please?

JBludau commented on core/src/impl/Kokkos_HostSpace_deepcopy.cpp:
```diff
@@ -137,10 +137,55 @@ void hostspace_parallel_deepcopy_async(const ExecutionSpace& exec, void* dst,
+      defined(KOKKOS_ENABLE_IMPL_HPX_ASYNC_DISPATCH))
+  constexpr uint64_t z_u64 = 0x0000000000000000;
+
+  // Align initial bytes to 8-byte boundary
+  uint8_t* dst_c = reinterpret_cast<uint8_t*>(dst);
+  size_t count   = 0;
+  while (reinterpret_cast<size_t>(dst_c) % 8 != 0) {
+    *dst_c = z_u8;
+    dst_c++;
+    count++;
+  }
+
+  // Zero-fill with 8-byte words in parallel
+  uint64_t* dst_p  = reinterpret_cast<uint64_t*>(dst_c);
+  const size_t cnt = (n - count) / 8;
+  Kokkos::parallel_for("Kokkos::Impl::hostspace_parallel_zeromemset",
+                       policy_t(exec, 0, cnt),
+                       [=](const size_t i) { dst_p[i] = z_u64; });
+
+  // Handle any remaining bytes that don't fit in 8-byte words
+  dst_c += cnt * 8;
+  uint8_t* dst_end = reinterpret_cast<uint8_t*>(dst) + n;
+  while (dst_c != dst_end) {
+    *dst_c = z_u8;
+    dst_c++;
+  }
```
hmm ... I am not sure if this might lead to interesting placements of the memory. Does iterating over the words produce a meaningful speedup or can the compiler autovectorize this?

Adrien-Tab replied:
I get a 10% speedup in the `PerfTest_ViewFirstTouch_Initialize.cpp` benchmark only. But the main reason for this change is that this part of the code was ridiculously slow in Debug mode.

JBludau replied:
I was just worried, that for more complex placements it might mess up the proximity. @crtrott do you think this could be an issue?

crtrott replied:
I think this is generally fine. The intention for default settings view is that each thread handles contiguous memory with parallel for. That's how the default iteration pattern MDRangePolicy over default Layout View ends up doing stuff for example.

JBludau commented on core/src/impl/Kokkos_HostSpace_deepcopy.hpp:
```diff
@@ -31,6 +31,9 @@ void hostspace_parallel_deepcopy_async(void* dst, const void* src, ptrdiff_t n);
 template <typename ExecutionSpace>
 void hostspace_parallel_deepcopy_async(const ExecutionSpace& exec, void* dst,
                                        const void* src, ptrdiff_t n);
+template <typename ExecutionSpace>
+void hostspace_parallel_zeromemset(const ExecutionSpace& exec, void* dst,
+                                   size_t n);
```
why is this in this header?

Adrien-Tab replied:
I couldn’t find a better place to put it

dalg24 replied:
`host_space_fence()` is also declared in that file so it's already a mixed bag.
Tidying up later is fine, I don't think we should block on this.

Adrien-Tab committed:
Add comments, rm dead code

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

dalg24 commented on core/perf_test/PerfTest_ViewFirstTouch_DeepCopy.cpp:
```diff
@@ -0,0 +1,42 @@
+  ViewType v_a("A", N);
+
+  for (auto _ : state) {
+    Kokkos::fence();
```
```suggestion
```

Adrien-Tab committed:
Remove fence

Signed-off-by: Adrien Taberner <adrien.taberner@outlook.com>

masterleinad commented on core/src/impl/Kokkos_HostSpace_deepcopy.cpp:
```diff
@@ -137,10 +137,55 @@ void hostspace_parallel_deepcopy_async(const ExecutionSpace& exec, void* dst,
 }
 
+template <typename ExecutionSpace>
+void hostspace_parallel_zeromemset(const ExecutionSpace& exec, void* dst,
+                                   size_t n) {
+  constexpr uint8_t z_u8 = 0x00;
+  using policy_t =
+      Kokkos::RangePolicy<ExecutionSpace, Kokkos::IndexType<size_t>>;
+
+#if !(defined(KOKKOS_ENABLE_HPX) && \
+      defined(KOKKOS_ENABLE_IMPL_HPX_ASYNC_DISPATCH))
+  constexpr uint64_t z_u64 = 0x0000000000000000;
```
Is this function ever called for `ExecutionSpace=Kokkos::Experimental::HPX`? Do we need the guards?

Adrien-Tab replied:
No, this function is never called for `HPX`. As For the guards, I simply followed the logic of what is written in the file.

masterleinad replied:
I'd remove it then. I mean it's weird that the implementation down below is different from what we are actually doing for HPX.

JBludau reviewed (approved):
lgtm but I would still be interested in @crtrott's answer concerning the hostspace zeromemset being chunked in the proposed way

crtrott commented on core/src/impl/Kokkos_HostSpace_deepcopy.cpp:
```diff
@@ -137,10 +137,55 @@ void hostspace_parallel_deepcopy_async(const ExecutionSpace& exec, void* dst,
 
+template void hostspace_parallel_zeromemset<DefaultHostExecutionSpace>(
+    const DefaultHostExecutionSpace&, void*, size_t);
+
```
Just a note: if we only enable Serial for the host, this instantiates a function which will never be called.

crtrott reviewed (approved):
We need to keep an eye on how much time we spend in benchmarking. However, this is not the worst offender right now.

dalg24 commented:
Please add to the 4.8 changelog
